### PR TITLE
Use Channel Instead of BufferBlock

### PR DESCRIPTION
**Pull request by jwood803**
Updates for #4482.

@eerhardt I took a shot at a single replacement to make sure it looks good before continuing. Does this look good or am I missing anything?

### CODE REVIEW

File: src/Microsoft.ML.Sweeper/AsyncSweeper.cs [hunk 1]
```diff
@@ -5,6 +5,7 @@
 using System;
 using System.Collections.Generic;
 using System.Threading;
+using System.Threading.Channels;
 using System.Threading.Tasks;
 using System.Threading.Tasks.Dataflow;
```
Comment (eerhardt) [inline src/Microsoft.ML.Sweeper/AsyncSweeper.cs:10]: We should be able to remove this using now.

File: src/Microsoft.ML.Sweeper/AsyncSweeper.cs [hunk 2]
```diff
@@ -168,7 +169,7 @@ public sealed class Options
         private readonly object _lock;
         private readonly CancellationTokenSource _cts;
 
-        private readonly BufferBlock<ParameterSetWithId> _paramQueue;
+        private readonly Channel<ParameterSetWithId> _paramQueue;
```
Comment (eerhardt) [inline src/Microsoft.ML.Sweeper/AsyncSweeper.cs:172]: What do you think about naming this `_paramChannel`?

File: src/Microsoft.ML.Sweeper/AsyncSweeper.cs [hunk 3]
```diff
@@ -208,7 +209,7 @@ public DeterministicSweeperAsync(IHostEnvironment env, Options options)
             _lock = new object();
             _results = new List<IRunResult>();
             _nullRuns = new HashSet<int>();
-            _paramQueue = new BufferBlock<ParameterSetWithId>();
+            _paramQueue = Channel.CreateUnbounded<ParameterSetWithId>();
```
Comment (eerhardt) [inline src/Microsoft.ML.Sweeper/AsyncSweeper.cs:212]: Are there any options we can set here? Specifically I think we can set [https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channeloptions.singlewriter?view=netcore-3.1](https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channeloptions.singlewriter?view=netcore-3.1)

Comment (jwood803) [reply src/Microsoft.ML.Sweeper/AsyncSweeper.cs:212]: Do you think there would be a single reader as well, or just a single writer?

Comment (eerhardt) [reply src/Microsoft.ML.Sweeper/AsyncSweeper.cs]: I haven't extensively looked through the code, but from my initial reading I don't think we could guarantee a single reader. Two threads could call `ProposeAsync()` at the same time.

The single writer is guaranteed by the `lock(_lock)` in the `public void Update(int id, IRunResult result)` method. (And the fact that the only other place `PrepareNextBatch` is called is from inside the constructor - which can't have multiple threads.)

**Comment by jwood803** [comment]
Thanks @eerhardt! I'll make similar changes to the other items and will mark this PR ready.

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs [hunk 4]
```diff
@@ -594,10 +594,10 @@ private async Task ProduceAsync()
                 try
                 {
                     int circularIndex = 0;
-                    while (await _toProduce.OutputAvailableAsync().ConfigureAwait(false))
+                    while (await _toProduceChannel.Reader.WaitToReadAsync().ConfigureAwait(false))
```
Comment (jwood803) [inline src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:597]: Not too sure if this is the best replacement for `OutputAvailableAsync` within the Channel.

File: src/Microsoft.ML.Data/Microsoft.ML.Data.csproj [hunk 5]
```diff
@@ -10,6 +10,7 @@
   <ItemGroup>
     <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonPackageVersion)" />
     <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
+    <PackageReference Include="System.Threading.Channels" Version="4.7.1" />
     <PackageReference Include="System.Threading.Tasks.Dataflow" Version="$(SystemThreadingTasksDataflowPackageVersion)" />
```
Comment (harishsk) [inline src/Microsoft.ML.Data/Microsoft.ML.Data.csproj:14]: We capture our repo wide dependency versions in build\Dependencies.props. Can you please add it there and use a named version like you see in the lines above and below?

Comment (jwood803) [reply src/Microsoft.ML.Data/Microsoft.ML.Data.csproj:14]: Updated. Thanks for catching this!

File: src/Microsoft.ML.Sweeper/Microsoft.ML.Sweeper.csproj [hunk 6]
```diff
@@ -6,6 +6,10 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="System.Threading.Channels" Version="$(SystemThreadingChannelsPackageVersion)" />
```
Comment (eerhardt) [inline src/Microsoft.ML.Sweeper/Microsoft.ML.Sweeper.csproj:10]: This shouldn't be needed because this project references `Microsoft.ML.Data`, which references this package.

File: src/Microsoft.ML.Data/Microsoft.ML.Data.csproj [hunk 7]
```diff
@@ -10,6 +10,7 @@
   <ItemGroup>
     <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonPackageVersion)" />
     <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
+    <PackageReference Include="System.Threading.Channels" Version="$(SystemThreadingChannelsPackageVersion)" />
     <PackageReference Include="System.Threading.Tasks.Dataflow" Version="$(SystemThreadingTasksDataflowPackageVersion)" />
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Microsoft.ML.Data.csproj:14]: We should be able to remove the dependency on `System.Threading.Tasks.Dataflow` now, right?

(Remove it here and in `build/Dependencies.props`)

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs [hunk 8]
```diff
@@ -559,28 +559,28 @@ protected override void Dispose(bool disposing)
...
-                    _toProduce.Complete();
+                    _toProduceChannel.Writer.Complete();
                     _producerTask.Wait();
 
                     // Complete the consumer after the producerTask has finished, since producerTask could
                     // have posted more items to _toConsume.
-                    _toConsume.Complete();
+                    _toConsumeChannel.Writer.Complete();
 
                     // Drain both BufferBlocks - this prevents what appears to be memory leaks when using the VS Debugger
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:569]: These comments need to be updated.

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs [hunk 9]
```diff
@@ -559,28 +559,28 @@ protected override void Dispose(bool disposing)
...
                     // because if a BufferBlock still contains items, its underlying Tasks are not getting completed.
                     // See https://github.com/dotnet/corefx/issues/30582 for the VS Debugger issue.
                     // See also https://github.com/dotnet/machinelearning/issues/4399
-                    _toProduce.TryReceiveAll(out _);
-                    _toConsume.TryReceiveAll(out _);
+                    _toProduceChannel.Reader.ReadAsync();
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:573]: We should test that this is still necessary. Read the linked issues for more information on why this was added. We might not need to drain these anymore with the move to Channel.

Comment (eerhardt) [reply src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:573]: If we do still need to drain these channels, this is wrong:

1. You need to synchronously read the item (you are doing `ReadAsync()` here and ignoring the result)
2. You need to read all of the messages, not just the first one

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs [hunk 10]
```diff
@@ -594,10 +594,10 @@ private async Task ProduceAsync()
...
                 {
                     int circularIndex = 0;
-                    while (await _toProduce.OutputAvailableAsync().ConfigureAwait(false))
+                    while (await _toProduceChannel.Reader.WaitToReadAsync().ConfigureAwait(false))
                     {
                         int requested;
-                        if (!_toProduce.TryReceive(out requested))
+                        if (!_toProduceChannel.Reader.TryRead(out requested))
                         {
                             // OutputAvailableAsync returned true, but TryReceive returned false -
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:602]: This comment needs to be updated.

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs [hunk 11]
```diff
@@ -652,14 +652,14 @@ protected override bool MoveNextCore()
...
-                    PostAssert(_toProduce, _deadCount);
+                    PostAssert(_toProduceChannel, _deadCount);
                     _deadCount = 0;
                 }
 
                 while (_liveCount < _poolRows && !_doneConsuming)
                 {
                     // We are under capacity. Try to get some more.
-                    int got = _toConsume.Receive();
+                    _toConsumeChannel.Reader.TryRead(out int got);
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:662]: You are ignoring the return value of `TryRead` here. It can return `false` if there is nothing to read right now.

File: build/Dependencies.props [hunk 12]
```diff
@@ -9,6 +9,7 @@
     <SystemMemoryVersion>4.5.1</SystemMemoryVersion>
     <SystemReflectionEmitLightweightPackageVersion>4.3.0</SystemReflectionEmitLightweightPackageVersion>
     <SystemThreadingTasksDataflowPackageVersion>4.8.0</SystemThreadingTasksDataflowPackageVersion>
+    <SystemThreadingChannelsPackageVersion>4.7.1</SystemThreadingChannelsPackageVersion>
```
Comment (harishsk) [inline build/Dependencies.props:12]: Can this be deleted now?

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs (hunk 9 above)
Comment (jwood803) [reply src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:573]: I did a very rudimentary test using a test console app that was in [#26565](https://github.com/dotnet/runtime/issues/26565) from @alandoherty. I saw the issue with `BufferBlock` but when changing to `Channels` I didn't see the issue. Even with and without completing the writer, the memory stayed consistent.

This may be safe to remove, but I'll double check by running the benchmark tests in a before and after to see if there are any big jumps in the numbers.

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs [hunk 13]
```diff
@@ -559,28 +559,28 @@ protected override void Dispose(bool disposing)
...
+                    // have posted more items to _toConsumeChannel.
+                    _toConsumeChannel.Writer.Complete();
 
                     // Drain both BufferBlocks - this prevents what appears to be memory leaks when using the VS Debugger
                     // because if a BufferBlock still contains items, its underlying Tasks are not getting completed.
                     // See https://github.com/dotnet/corefx/issues/30582 for the VS Debugger issue.
                     // See also https://github.com/dotnet/machinelearning/issues/4399
-                    _toProduce.TryReceiveAll(out _);
-                    _toConsume.TryReceiveAll(out _);
+                    //_toProduceChannel.Reader.ReadAsync();
```
Comment (jwood803) [inline src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:573]: This was left in for performance tests. Will report on what I find.

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs (hunk 9 above)
Comment (jwood803) [reply src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs]: @eerhardt Ran a couple of benchmarks and below are some results. It did complain that I didn't have it compiled in Release mode even though I ran the build script with it and built in Release in Visual Studio so these results may be under Debug mode.

They look very similar with and without having to flush out the Channels. I'm thinking this may be safe to remove but curious about your thoughts.

Without update:
|            Method |   Mean |  Error | StdDev | Extra Metric |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|------------------- |--------:|--------:|--------:|-------------:|-------------:|------------:|-----------:|----------:|
| TrainFeaturizeText | 23.30 s | 23.88 s | 1.309 s |           - | 1997000.0000 | 781000.0000 | 40000.0000 | 320.3 MB |
 
|         Method |   Mean |   Error | StdDev | Extra Metric |     Gen 0 |     Gen 1 |     Gen 2 | Allocated |
|---------------- |--------:|---------:|--------:|-------------:|-----------:|-----------:|-----------:|----------:|
| TrainResnetV250 | 67.38 s | 283.41 s | 15.53 s |           - | 75000.0000 | 34000.0000 | 11000.0000 |  2.01 GB |
 
With update:
|            Method |   Mean |  Error | StdDev | Extra Metric |       Gen 0 |      Gen 1 |     Gen 2 | Allocated |
|------------------- |--------:|--------:|--------:|-------------:|-------------:|------------:|-----------:|----------:|
| TrainFeaturizeText | 19.77 s | 29.54 s | 1.619 s |           - | 1978000.0000 | 771000.0000 | 34000.0000 | 320.3 MB |
 
|         Method |   Mean |   Error | StdDev | Extra Metric |     Gen 0 |     Gen 1 |     Gen 2 | Allocated |
|---------------- |--------:|---------:|--------:|-------------:|-----------:|-----------:|-----------:|----------:|
| TrainResnetV250 | 54.75 s | 111.72 s | 6.124 s |           - | 76000.0000 | 34000.0000 | 11000.0000 |  2.01 GB |

File: src/Microsoft.ML.Data/Microsoft.ML.Data.csproj (hunk 7 above)
Comment (jwood803) [reply src/Microsoft.ML.Data/Microsoft.ML.Data.csproj:14]: I think this is still needed. The build yelled when I had it removed 😄

Comment (eerhardt) [reply src/Microsoft.ML.Data/Microsoft.ML.Data.csproj:14]: What was the error? From what I remember `BufferBlock` was the only thing we were using from `Dataflow`.

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs (hunk 9 above)
Comment (eerhardt) [reply src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs]: > I saw the issue with BufferBlock but when changing to Channels I didn't see the issue. Even with and without completing the writer, the memory stayed consistent.

Then I think we are OK to remove this "draining" code altogether.

File: src/Microsoft.ML.Data/Microsoft.ML.Data.csproj (hunk 7 above)
Comment (jwood803) [reply src/Microsoft.ML.Data/Microsoft.ML.Data.csproj:14]: Actually, I misread the error. It seems that it was still trying to restore the package and since there wasn't a version defined it failed. Latest push should fix that.

File: pkg/Microsoft.ML/Microsoft.ML.nupkgproj [hunk 14]
```diff
@@ -11,7 +11,6 @@
 
     <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonPackageVersion)" />
     <PackageReference Include="System.Reflection.Emit.Lightweight" Version="$(SystemReflectionEmitLightweightPackageVersion)" />
-    <PackageReference Include="System.Threading.Tasks.Dataflow" Version="$(SystemThreadingTasksDataflowPackageVersion)" />
```
Comment (eerhardt) [inline pkg/Microsoft.ML/Microsoft.ML.nupkgproj:14]: You also need to add `<PackageReference Include="System.Threading.Channels" Version="$(SystemThreadingChannelsPackageVersion)" />` here.

**Comment by eerhardt** [comment]
I think you can remove this line as well:

https://github.com/dotnet/machinelearning/blob/665a3669652bc478fffbc7ed73b89067e5a7938c/test/Microsoft.ML.FSharp.Tests/SmokeTests.fs#L25

File: src/Microsoft.ML.Sweeper/Microsoft.ML.Sweeper.csproj [hunk 15]
```diff
@@ -5,7 +5,7 @@
     <DefineConstants>CORECLR</DefineConstants>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
   </PropertyGroup>
-
+
```
Comment (eerhardt) [inline src/Microsoft.ML.Sweeper/Microsoft.ML.Sweeper.csproj:8]: Can you revert this change?

**Comment by eerhardt** [comment]
>             protected override void Dispose(bool disposing)

This whole `Dispose` method looks like it can be deleted now.

---
Refers to: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:555 in 8a45c1d. [](commit_id = 8a45c1d8e1a707a81e6acd9b83c6c5c5c2f5fa26, deletion_comment = False)

File: src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs [hunk 16]
```diff
@@ -651,26 +635,29 @@ protected override bool MoveNextCore()
...
+                    PostAssert(_toProduceChannel, _deadCount);
                     _deadCount = 0;
                 }
 
                 while (_liveCount < _poolRows && !_doneConsuming)
                 {
                     // We are under capacity. Try to get some more.
-                    int got = _toConsume.Receive();
-                    if (got == 0)
+                    var hasReadItem = _toConsumeChannel.Reader.TryRead(out int got);
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Transforms/RowShufflingTransformer.cs:646]: Previously `_toConsume.Receive()` would block the thread until something was available. But now `TryRead` will return `false` immediately if something wasn't available. In that case, we will keep spinning in the `while` loop and checking the reader constantly.

It might make sense to keep the previous behavior and block the thread until `_toConsumeChannel` has an item available, so we aren't wasting CPU cycles spinning here.

**Comment by eerhardt** [comment]
> Is there a reason why we haven't yet merged in this PR?

I don't believe this comment has been addressed yet:

https://github.com/dotnet/machinelearning/pull/5123#discussion_r434105449

**Comment by jwood803** [comment]
> I don't believe this comment has been addressed yet

@eerhardt I may need some guidance on the best way to block the thread there. 😃 Would using the `lock` keyword work for that?

**Comment by eerhardt** [comment]
I believe you can call [`ChannelReader<T>.WaitToRead`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channelreader-1.waittoreadasync?view=netcore-3.1) and block on it. This is similar to how `Receive()` works:

https://github.com/dotnet/runtime/blob/4cb924de165aa8e5373a109a68611d422aa72a09/src/libraries/System.Threading.Tasks.Dataflow/src/Base/DataflowBlock.cs#L952-L959

```C#
            // Get a TCS to represent the receive operation and wait for it to complete.
            // If it completes successfully, return the result. Otherwise, throw the
            // original inner exception representing the cause.  This could be an OCE.
            Task<TOutput> task = ReceiveCore(source, false, timeout, cancellationToken);
            try
            {
                return task.GetAwaiter().GetResult(); // block until the result is available
            }
```

**Comment by jwood803** [comment]
@eerhardt Updated to use `WaitToReadAsync`, but definitely let me know if more changes are needed.

**Comment by harishsk** [comment]
@jwood803 Thank you for your contribution. I am merging the PR now.
